src/System.Private.ServiceModel/tests/Scenarios/Client/ExpectedExceptions/ExpectedExceptionTests.cs

```diff
@@ -248,6 +247,144 @@ public static void FaultException_Throws_WithFaultDetail()
+    }
+
+    [Theory]
+    [OuterLoop]
+    public static void FaultException_Throws_FaultDetail2()
+    {
+        string faultMsg = "Test Fault Exception";
+        StringBuilder errorBuilder = new StringBuilder();
+
+        try
+        {
+            BasicHttpBinding binding = new BasicHttpBinding();
+            using (ChannelFactory<IWcfService> factory = new ChannelFactory<IWcfService>(binding, new EndpointAddress(Endpoints.HttpBaseAddress_Basic)))
+            {
+                IWcfService serviceProxy = factory.CreateChannel();
+                serviceProxy.TestFaults(faultMsg, false);
+            }
+        }
+        catch (Exception e)
+        {
+            if (e.GetType() != typeof(FaultException<FaultDetail2>))
+            {
+                string error = string.Format("Expected exception: {0}, actual: {1}\r\n{2}",
+                                             "FaultException<FaultDetail2>", e.GetType(), e.ToString());
+                if (e.InnerException != null)
+                    error += String.Format("\r\nInnerException:\r\n{0}", e.InnerException.ToString());
+                errorBuilder.AppendLine(error);
+            }
+            else
+            {
+                FaultException<FaultDetail2> faultException = (FaultException<FaultDetail2>)(e);
+                string actualFaultMsg = ((FaultDetail2)(faultException.Detail)).Message;
+                if (actualFaultMsg != faultMsg)
+                {
+                    errorBuilder.AppendLine(string.Format("Expected Fault Message: {0}, actual: {1}", faultMsg, actualFaultMsg));
+                }
+            }
+
+            Assert.True(errorBuilder.Length == 0, string.Format("Test Scenario: FaultException_Throws_WithFaultDetail2 FAILED with the following errors: {0}", errorBuilder));
+            return;
+        }
+
+        Assert.True(false, "Expected FaultException<FaultDetail2> exception, but no exception thrown.");
+    }
+
+    [Theory]
+    [OuterLoop]
+    public static void FaultException_Echo_Or_Throws_Fault_Exception()
+    {
+        string faultMsg = "Test Fault Exception";
+        StringBuilder errorBuilder = new StringBuilder();
+
+        try
+        {
+            BasicHttpBinding binding = new BasicHttpBinding();
+            using (ChannelFactory<IWcfService> factory = new ChannelFactory<IWcfService>(binding, new EndpointAddress(Endpoints.HttpBaseAddress_Basic)))
+            {
+                var obj = new FaultDetail2("FaultDetail2 message");
+                var input = new object[] { obj };
+
+                IWcfService serviceProxy = factory.CreateChannel();
+                var response = serviceProxy.TestFaultWithKnownType(null, input);
+
+                Assert.Equal(typeof(FaultDetail2), response[0].GetType());
+                Assert.Equal(obj.Message, ((FaultDetail2)response[0]).Message);
+
+                // Expect exception
+                serviceProxy.TestFaultWithKnownType(faultMsg, null);
```

Comment:
I've split this into FaultException_FaultContractAndKnownType_Echo() and FaultException_FaultContractAndKnownType_Throws_WithFaultDetail()